identity_credential/src/validator/domain_linkage_validator.rs

```diff
@@ -0,0 +1,530 @@
+use crate::validator::FailFast;
+use identity_core::common::OneOrMany;
+use identity_core::common::Url;
+use identity_did::CoreDID;
+use serde::Serialize;
+use std::collections::HashSet;
+
+type DomainLinkageValidationResult = Result<(), DomainLinkageValidationError>;
+
+/// A validator for a Domain Linkage Configuration and Credentials.
+pub struct DomainLinkageValidator {}
+
+impl DomainLinkageValidator {
+  /// Validates the linkage between a domain and a DID.
+  /// [`DomainLinkageConfiguration`] is validated according to [DID Configuration Resource Verification](https://identity.foundation/.well-known/resources/did-configuration/#did-configuration-resource-verification).
+  ///
+  /// * `issuer`: DID Document of the linked DID. Issuer of the Domain Linkage Credential included
+  /// in the Domain Linkage Configuration.
+  /// * `configuration`: Domain Linkage Configuration fetched from the domain at "/.well-known/did-configuration.json".
+  /// * `domain`: domain from which the Domain Linkage Configuration has been fetched.
+  /// * `validation_options`: Further validation options to be applied on the Domain Linkage Credential.
+  ///
+  /// # Note:
+  /// - Only [Linked Data Proof Format](https://identity.foundation/.well-known/resources/did-configuration/#linked-data-proof-format)
+  ///   is supported.
+  /// - Only the Credential issued by `issuer` is verified.
+  ///
+  /// # Errors
+  ///  - Semantic structure of `configuration` is invalid.
+  ///  - `configuration` includes multiple credentials issued by `issuer`.
+  ///  - Validation of the matched Domain Linkage Credential fails.
+  pub fn validate_linkage<DOC: ValidatorDocument>(
+    issuer: &DOC,
+    configuration: &DomainLinkageConfiguration,
+    domain: &Url,
+    validation_options: &CredentialValidationOptions,
+  ) -> DomainLinkageValidationResult {
+    let mut matched_credentials = configuration
+      .linked_dids()
+      .iter()
+      .filter(|credential| credential.issuer.url().as_str() == issuer.did_str());
+
+    match matched_credentials.next() {
+      None => Err(DomainLinkageValidationError {
+        cause: DomainLinkageValidationErrorCause::InvalidStructure,
+        source: None,
+      }),
+      Some(credential) => {
+        if matched_credentials.next().is_some() {
+          Err(DomainLinkageValidationError {
+            cause: DomainLinkageValidationErrorCause::InvalidStructure,
+            source: None,
+          })
+        } else {
+          Self::validate_credential(issuer, credential, domain.clone(), validation_options)
+        }
+      }
+    }
+  }
+
+  /// Validates a [Domain Linkage Credential](https://identity.foundation/.well-known/resources/did-configuration/#domain-linkage-credential).
+  ///
+  /// *`issuer`: issuer of the credential.
+  /// *`credential`: domain linkage Credential to be verified.
+  /// *`domain`: the domain hosting the credential.
+  pub fn validate_credential<T: Serialize, DOC: ValidatorDocument>(
```

Comment:
Would be great if you could add some comments in this functions to each section of the validation. Because the function is so long it would be helpful to have a rough overview of what each section validates.
Alternatively, split the function into smaller functions with descriptive names.
(Or whatever you discuss with Oliver).